### DISCUSSION TITLE
Any way I can use MistralAi's Codestral through their API?

### DISCUSSION THREAD

**Discussion by frodew** [olimorris/codecompanion.nvim#282] [Q&A] [answered]
I am not really familiar on how this works so I wondered whether someone could help me out here: Seems like a really cool plugin to neovim, but I do not really want to use the bigger LLMs. I looked around and something smaller like MistralAI's codestral seems to work quite fine. My laptop is to weak to run ollama locally and deploying it to a server seems to expensive and an overkill for my occasional usage.
So I guess my best option would be to go through Mistrals API, but I don't know whether this works with this plugin. Or is there an option to let it "run through ollama somehow"?

**Comment by olimorris** [comment, marked as answer]
The [adapters](https://github.com/olimorris/codecompanion.nvim/blob/main/doc/ADAPTERS.md) guide should have you covered. Also #279 is a recent resource and an example of how you can create your own adapter to work with the plugin.

**Reply by frodew** [reply]
Perfect - exactly what I was looking for; thank you!

**Comment by kronolynx** [comment]
This is what worked for me

```lua
{
    "olimorris/codecompanion.nvim",
    dependencies = {
      "nvim-lua/plenary.nvim",
      "nvim-treesitter/nvim-treesitter",
      -- The following are optional:
      { "MeanderingProgrammer/render-markdown.nvim", ft = { "markdown", "codecompanion" } },
    },
    config = function()
      require("codecompanion").setup({
        strategies = {
          chat = {
            adapter = "openai",
          },
          inline = {
            adapter = "openai",
          },
        },
        opts = {
          stream = true,
        },
        adapters = {
          openai = function()
            return require("codecompanion.adapters").extend("openai_compatible", {
              name = "codestral",
              env = {
                url = "https://api.mistral.ai",
                api_key = "CODESTRAL_API_KEY",
                chat_url = "/v1/chat/completions"
              },
              handlers = {
                form_parameters = function(self, params, messages)
                  -- codestral doesn't support these in the body
                  params.stream_options = nil
                  params.options = nil

                  return params
                end,
              },
              schema = {
                model = {
                  default = "codestral-latest",
                },
                temperature = {
                  default = 0.2,
                  mapping = "parameters", -- not supported in default parameters.options
                },
              },
            })
          end,
        },
      })
    end
  }
```

**Comment by rmsc** [comment]
I have made a custom adapter for the mistral API, using the same approach as the xAI adapter. Would you be open to a PR?

**Reply by flovilmart** [reply]
Exactly when I started to want to write my own!

**Reply by rmsc** [reply]
A nice coincidence I guess :)

There's a PR under review: https://github.com/olimorris/codecompanion.nvim/pull/1118
